openedx/features/course_card/views.py

```diff
@@ -48,13 +50,25 @@ def get_course_cards(request):
         if course.invitation_only and not CourseEnrollment.is_enrolled(request.user, course.id):
             continue
 
+        programs = get_programs(request.site)
+
+        # check if any active (currently open or are starting in future)
+        # run of a course is part of a program.
+        course_program = [program for program in programs
+                          for program_course in program['courses']
+                          if program_course['course_runs']
+                          and str(course.id) in [rerun['key'] for rerun in program_course['course_runs']
+                                                 if rerun['enrollment_start'] and rerun['enrollment_end']
+                                                 and (date_from_str(rerun['enrollment_start']) > current_time
+                                                 or date_from_str(rerun['enrollment_start']) <= current_time <= date_from_str(rerun['enrollment_end']))]]
```

Comment:
this way the line is more readable. I tried breaking it down but that makes it more difficult to comprehend

Reply:
This portion of code is too complicated with high time complexity. It took me some time to understand what it is exactly doing, as pointed out by @ahmed-faraz46. Therefore, I've come up with some new implementation that divides the portion of code into multiple functions and also reduces time complexity:

First of all, such complex code should not be introduced in the existing functions, rather a new function should be created for it.

# replace existing code (line 54-62) with this
```
        if is_course_part_of_specialization(course.id, programs, current_time):
            course.specialization = True
```

# new methods

```
def is_course_part_of_specialization(course_id, programs, current_time):
    for program in programs:
        for program_course in program['courses']:
            if program_course['course_runs']:
                for rerun in program_course['course_runs']:
                    if is_course_run_active(rerun, current_time) and course_id == rerun:
                        return True
    return False


def is_course_run_active(rerun, current_time):
    if rerun['enrollment_end'] and current_time <= date_from_str(rerun['enrollment_end']):
        return True
    return False

```

Reply:
Justification in favour of new implementation:

1. the new method 'is_course_run_active' more efficiently determines if a course run is active. According to previous approach, we had multiple conditionals, the need for those conditionals is eliminated in my opinion if we use this conditional only. It covers both the scenarios. If we know that the enrolment end date of a course run hasn't reached yet, it automatically implies that the course run is active. In simpler terms, the previous conditions seemed redundant.

3. the method 'is_course_part_of_specialization' is much more efficient because all we need in the output is a boolean, i.e. whether a course is part of a specialization program or not. There is no need to first get all course programs against a particular course. The moment we hit an active course run against our course id, that's it, we return true because we would add the specialisation tag to that course, if I'm not wrong. Similarly, there is no need to get all course run either, we should use the exit first approach, the moment we find what is required, break the loop/exit from the function. This will help reduce unnecessary computation of getting a list of all reruns and course programs everytime.

Reply:
And @nofalfayyaz I have some other questions as well regarding the existing approach. We (@ahmed-faraz46, you and I) can discuss it in call. And @ahmed-faraz46 please correct me if there is something wrong with my suggested implementation. I have just modified the existing implementation, without changing much logic or approach. If there is altogether a different approach like you mentioned of using dictionary/not updating model, then we'll obviously go with that. But I think the existing implementation would still help in identifying courses belonging to specialisation programs.

Reply:
Sure, Let's please schedule a call on Monday.